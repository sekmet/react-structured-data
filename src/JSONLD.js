import React from "react";
import PropTypes from "prop-types";
import GenericNodeCollection from "./core/GenericNodeCollection";

export const JSONLD = props => {
  let json = null;
  if (props.children) {
    const ChildClass = props.children.type;
    const { children, type, ...schema } = props.children.props;

    const childClassInstance = new ChildClass(props.children.props);
    const firstChild = childClassInstance.getJSON(schema);

    json =
      childClassInstance instanceof GenericNodeCollection
        ? Object.assign(
            { "@context": "https://schema.org/" },
            { [type]: firstChild }
          )
        : Object.assign({ "@context": "https://schema.org/" }, firstChild);
  }

  return (
    <script
      type="application/ld+json"
      dangerouslySetInnerHTML={{ __html: JSON.stringify(json) }}
    />
  );
};

JSONLD.propTypes = {
  additionalType: PropTypes.object,
  children: PropTypes.node
};

export default JSONLD;
