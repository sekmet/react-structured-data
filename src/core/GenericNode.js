import JSONLDAbstractNode from "../JSONLDAbstractNode";
import { encode } from "he";

class GenericNode extends JSONLDAbstractNode {
  getJSON({ type, jsonldtype, ...schema }) {
    const parseChildren = super.parseChildren();
    const details = {
      "@type": jsonldtype || this.getSchemaOrgType(),
      ...schema
    };
    // delete props that are `null`
    Object.keys(details).forEach(key => {
      if (details[key] === null) delete details[key];
    });
    // escape string prop values
    const clearProps = Object.assign(
      ...Object.entries(details).map(([key, value]) => ({
        [key]: typeof value === "string" ? encode(value) : value
      }))
    );
    return type
      ? Object.assign({ [type]: clearProps }, ...parseChildren)
      : Object.assign(clearProps, ...parseChildren);
  }
}

export default GenericNode;
