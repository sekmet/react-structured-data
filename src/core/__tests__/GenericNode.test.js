import React from "react";
import renderer from "react-test-renderer";
import JSONLD from "../../JSONLD";
import GenericNode from "../GenericNode";

describe("GenerticNode", () => {
  it("renders question to QAPage mainEntity as Question type", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNode
            type="review"
            jsonldtype="Review"
            name="It is awesome"
            reviewBody="This is great!"
          >
            <GenericNode
              type="itemReviewed"
              jsonldtype="Product"
              id="product-x"
            />
            <GenericNode type="author" jsonldtype="Person" name="Cool Carl" />
            <GenericNode
              type="locationCreated"
              jsonldtype="AdministrativeArea"
              name="Chicago, IL"
            />
          </GenericNode>
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
  it("filters out null and false child nodes", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNode
            type="review"
            jsonldtype="Review"
            name="It is awesome"
            reviewBody="This is great!"
          >
            <GenericNode
              type="itemReviewed"
              jsonldtype="Product"
              id="product-x"
            />
            {null}
            {false}
            <GenericNode
              type="locationCreated"
              jsonldtype="AdministrativeArea"
              name="Chicago, IL"
            />
          </GenericNode>
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
  it("filters out properties that have null values", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNode
            type="review"
            jsonldtype="Review"
            name="It is awesome"
            reviewBody={null}
          />
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
  it("filters out one child which is null", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNode
            type="review"
            jsonldtype="Review"
            name="It is awesome"
            reviewBody={null}
          >
            {null}
          </GenericNode>
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
  it("filters out one child which is false", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNode
            type="review"
            jsonldtype="Review"
            name="It is awesome"
            reviewBody={null}
          >
            {false}
          </GenericNode>
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
  it("escapes json xss", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNode
            type="review"
            jsonldtype="Review"
            reviewBody={`</script><script>alert('XSS');</script>`}
          />
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
  it("escapes javascript xss", () => {
    expect(
      renderer.create(
        <JSONLD>
          <GenericNode
            type="review"
            jsonldtype="Review"
            reviewBody={`\\";alert('XSS');//`}
          />
        </JSONLD>
      )
    ).toMatchSnapshot();
  });
});
