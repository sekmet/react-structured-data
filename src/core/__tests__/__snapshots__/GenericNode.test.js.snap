// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerticNode escapes javascript xss 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Review\\",\\"reviewBody\\":\\"\\\\\\\\&#x22;;alert(&#x27;XSS&#x27;);//\\"}",
    }
  }
  type="application/ld+json"
/>
`;

exports[`GenerticNode escapes json xss 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Review\\",\\"reviewBody\\":\\"&#x3C;/script&#x3E;&#x3C;script&#x3E;alert(&#x27;XSS&#x27;);&#x3C;/script&#x3E;\\"}",
    }
  }
  type="application/ld+json"
/>
`;

exports[`GenerticNode filters out null and false child nodes 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Review\\",\\"name\\":\\"It is awesome\\",\\"reviewBody\\":\\"This is great!\\",\\"itemReviewed\\":{\\"@type\\":\\"Product\\",\\"@id\\":\\"product-x\\"},\\"locationCreated\\":{\\"@type\\":\\"AdministrativeArea\\",\\"name\\":\\"Chicago, IL\\"}}",
    }
  }
  type="application/ld+json"
/>
`;

exports[`GenerticNode filters out one child which is false 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Review\\",\\"name\\":\\"It is awesome\\"}",
    }
  }
  type="application/ld+json"
/>
`;

exports[`GenerticNode filters out one child which is null 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Review\\",\\"name\\":\\"It is awesome\\"}",
    }
  }
  type="application/ld+json"
/>
`;

exports[`GenerticNode filters out properties that have null values 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Review\\",\\"name\\":\\"It is awesome\\"}",
    }
  }
  type="application/ld+json"
/>
`;

exports[`GenerticNode renders question to QAPage mainEntity as Question type 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Review\\",\\"name\\":\\"It is awesome\\",\\"reviewBody\\":\\"This is great!\\",\\"itemReviewed\\":{\\"@type\\":\\"Product\\",\\"@id\\":\\"product-x\\"},\\"author\\":{\\"@type\\":\\"Person\\",\\"name\\":\\"Cool Carl\\"},\\"locationCreated\\":{\\"@type\\":\\"AdministrativeArea\\",\\"name\\":\\"Chicago, IL\\"}}",
    }
  }
  type="application/ld+json"
/>
`;
